### ISSUE TITLE
make aux type more generic

### ISSUE THREAD

**Issue by zpgaal** [omni-viral/rendy#49]
As I see the provided aux type is never used by the graph. 
Can we make it more general to let the client decide if it is &mut T, or &T or a "moved" T object ?

In other word make the &mut part of the generic argument:
Graph<Backend, &mut World> instead of Graph<Backend, World>
or Graph<Backend, &World> instead of Graph<Backend, World>
build(aux: T) instead of build(aux: &mut T)

**Comment by omni-viral** [comment]
@zpgaal I was thinking about it.
Lets assume that you have created `GraphBuilder<B, &'a mut T>`.
Now you call `GraphBuilder<B, &'a mut T>::build` and this call borrows `T` for entire `'a`.
Now you can't call `Graph<B, &'a mut T>::run` because `T` is borrowed.
Actually even `GraphBuilder<B, &'a mut T>::build` wont compile because it tries to give `&'a mut T` to each node builder, but `&'a mut T` is not clonable and reborrowing doesn't work because it produces lesser lifetime than node requires.

**Comment by zpgaal** [comment]
Ok i'll try to do some test without any promise :)

Another option could be to make T:Clone and for the current behavior one have to use some kind  internal mutability (refcell, arc, mutex, etc.)

Some background info: I' m experimenting with shred (as i remember amethyst also relies on it). And it has its own guards for mut and non-mut handling (like mutex). For rendy I had to encapsulate the resources (world) not to require a mutable capture of the whole state. Though i'm not quite sure how will the update of the systems in shred and graph traverse work together.

**Comment by omni-viral** [comment]
If you take a look at `split` branch you'll see `Graph::run` method will take `aux: &T`.

So you will only need mutable access during graph building. I would like to relax that requirement as well, but not figured yet how to do that.

**Comment by zpgaal** [comment]
I've tried make the T an argument of the functions, but of course it wont compile (cannot be made into object). I've also tried the Rc<RefCell>/Mutex/Arc combos, it could work but I would not say it is any better.
Another issue is that T must be 'static that prevents to pass any structure with reference.

I think GAT might be a solution, but that's not an option at the moment:
```
trait Graph<B> {
  type Aux<'a>;
  fn run<'a>(aux: Aux<'a>);
}
```

But for now I'm out of ideas that does not involve unsafe.
